byron/semantics/executable-spec/src/Control/State/Transition/Trace.hs

```diff
@@ -346,3 +347,20 @@ extractValues d =  catMaybes (gmapQ extractValue d)
   where
     extractValue :: forall d1 . (Data d1) => d1 -> Maybe a
     extractValue d1 = cast d1
+
+
+-- | Extract triplets of the form [(s, sig, t)] from a trace. For a valid trace,
+-- each source state can reach a target state via the given signal.
+--
+sourceSignalTargets :: forall a. Trace a -> [(State a, Signal a, State a)]
+sourceSignalTargets (Trace _ _ []) = [] -- only init state exists
+sourceSignalTargets (Trace _ sInit ts) =
+  doIt sInit (reverse ts) []
```

Comment:
Here we might want to use the `preStatesAndSignals` to avoid duplication of logic. If we would change the representation of trace we would need to change this function as well. I'm thinking that it might be a good idea to move `Trace` to an `Internal` module.

Also I'm wondering whether this function is equivalent to something like:

```haskell
sourceSignalTargets trace = zip3 states signals (tail states)
where 
  states = traceStates OldestFirst trace
  signals = traceSignals OldestFirst trace
```